### PR TITLE
SQS: Add send a message tab

### PR DESCRIPTION
**Pull request by ranyoo2367**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Added a tab for sending message. 
- FIFO Queues:
   - Asks for message body, deduplication ID, and group ID.
   - If input fields are empty upon pressing send, error messages are displayed.
   - Context help added for deduplication and group IDs.
- Standard Queues:
   - Asks for message body.
   - If message body is empty upon pressing send, error message is displayed.
- A confirmation with the message ID is displayed if message is sent successfully. If it fails to send the message, an error message is displayed.

## Testing

- Tested form validation to make sure error messages are displayed if any of the message body, deduplication ID, or group ID fields are empty.
- Tested confirmation/failure of sending message.

## Screenshots (if appropriate)
FIFO:
FifoEdit

Standard:
Screen Shot 2020-08-04 at 9 57 28 PM

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by hunterwerlla** [comment]
If I make the message too long, the box can't scroll and the send button dissapears
Screen Shot 2020-08-03 at 2 47 31 PM

**Comment by hunterwerlla** [comment]
I don't think there needs to be a separator in this menu:
Screen Shot 2020-08-03 at 2 55 23 PM

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt [hunk 1]
```diff
@@ -2,8 +2,119 @@
...
+    }
+
+    private fun validateFields(): Boolean {
+        emptyBodyLabel.isVisible = inputText.text.isEmpty()
+        if (isFifo) {
+            emptyDeduplicationLabel.isVisible = deduplicationId.text.isEmpty()
+            emptyGroupLabel.isVisible = groupId.text.isEmpty()
+        }
+
+        return !(emptyBodyLabel.isVisible || emptyDeduplicationLabel.isVisible || emptyGroupLabel.isVisible)
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt:118]: This check should be based on what isVisible is based on, not the state of isVisible. You can save the isEmpty checks to booleans then reuse them

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt [hunk 2]
```diff
@@ -2,8 +2,119 @@
...
 import javax.swing.JPanel
 
-class SendMessagePane {
+class SendMessagePane(
+    private val client: SqsClient,
+    private val queue: Queue
+) : CoroutineScope by ApplicationThreadPoolScope("SendMessagePane") {
     lateinit var component: JPanel
+    lateinit var inputText: JBTextArea
+    lateinit var fifoComponent: JPanel
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.kt:28]: `fifoComponent` does not really explain what it is. Maybe `fifoFields` or `fifoArguments` or something to that effect

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form [hunk 3]
```diff
@@ -1,12 +1,155 @@
...
+    <children>
+      <component id="94dd3" class="javax.swing.JLabel">
+        <constraints>
+          <grid row="0" column="1" row-span="1" col-span="3" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Message Body:"/>
+        </properties>
+      </component>
+      <grid id="64b6c" binding="fifoComponent" layout-manager="GridLayoutManager" row-count="4" column-count="4" same-size-horizontally="false" same-size-vertically="false" hgap="-1" vgap="-1">
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:19]: consider making fifoComponent its own form, it has as many components as the other form, and can help encapsulate the logic to have less checks in the main form.

**Comment by hunterwerlla** [comment]
FIFO queues by default have a tiny message window:
Screen Shot 2020-08-03 at 3 08 10 PM

**Comment by hunterwerlla** [comment]
Sending a FIFO message with the same deduplication ID and group ID fails silently

**Comment by hunterwerlla** [comment]
A general UX thing, as soon as the deduplication/group ID box has text in it, we can hide the messages again

**Comment by hunterwerlla** [comment]
If I put in a dedupliction ID that's too long, I get a generic "Failed to send message", we should be able to bubble up this error

**Comment by ranyoo2367** [comment]
New set of changes:
- Fifo components are in a separate form.
- Replaced `PersistentStateComponent` with `PropertiesComponent`
- Added listeners for components
- Added more specific validation errors
- UI changes to enable scrolling for text area

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form [hunk 4]
```diff
@@ -1,12 +1,90 @@
...
+        <constraints>
+          <grid row="4" column="2" row-span="1" col-span="1" vsize-policy="1" hsize-policy="6" anchor="0" fill="1" indent="0" use-parent-layout="false"/>
+        </constraints>
+      </hspacer>
+      <component id="6d3df" class="javax.swing.JButton" binding="sendButton">
+        <constraints>
+          <grid row="4" column="3" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="2" fill="1" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Send"/>
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:69]: Needs to be localizable

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form [hunk 5]
```diff
@@ -1,12 +1,90 @@
...
+        </properties>
+      </component>
+      <component id="7165c" class="javax.swing.JButton" binding="clearButton">
+        <constraints>
+          <grid row="4" column="4" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="2" fill="1" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Clear"/>
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePane.form:77]: Ditto

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @Test
+    fun `No input fails to send for standard`() {
+        standardPane.apply {
+            inputText.text = ""
+            runBlocking { sendMessage() }
+        }
+
+        assertTrue { standardPane.bodyErrorLabel.isVisible }
```
Comment (hunterwerlla) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt:54]: Ah I didn't notice this before, but we don't use kotlin test asserts (I need to add that to the import ktlint rule), we use the asssertJ versions

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            runBlocking { sendMessage() }
+        }
+
+        assertFalse { fifoPane.bodyErrorLabel.isVisible }
+        assertFalse { fifoPane.fifoFields.deduplicationErrorLabel.isVisible }
+        assertTrue { fifoPane.fifoFields.groupErrorLabel.isVisible }
+    }
+
+    @Test
+    fun `Too long ID fails to send for fifo`() {
```
Comment (hunterwerlla) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/toolwindow/SendMessagePaneTest.kt:101]: nit: `ID too long`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            document.addDocumentListener(createListener(this, groupErrorLabel))
+        }
+    }
+
+    fun validateFields(): Boolean {
+        var deduplicationIsValid = true
+        var groupIsValid = true
+
+        if (deduplicationId.text.length > MAX_LENGTH_OF_FIFO_ID) {
+            deduplicationIsValid = false
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt:62]: Add a message for this then put the next if in an `else if`?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        if (deduplicationId.text.isEmpty()) {
+            deduplicationErrorLabel.text = message("sqs.message.validation.empty.deduplication_id")
+            deduplicationErrorLabel.isVisible = true
+            deduplicationIsValid = false
+        }
+        if (groupId.text.length > MAX_LENGTH_OF_FIFO_ID) {
```
Comment (hunterwerlla) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt:69]: Ditto

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt (hunk 8 above)
Comment (ranyoo2367) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/toolwindow/FifoPanel.kt:62]: The message should be taken care of in the document listener, but I can explicitly add it again.

**Pull request merged**